### PR TITLE
Add profiling

### PR DESCRIPTION
**Pull request by SomeKittens**
This fixes a previous performance issue and emits an event after each digest with details on the length of the digest as well as the length of each watcher.

https://github.com/angular/angular-hint/pull/76 needs to be merged in ahead of this in order for the example page to work.

### CODE REVIEW

File: src/modules/scopes.js [hunk 1]
```diff
@@ -72,63 +72,74 @@ function decorateRootScope($delegate, $parse) {
...
+        var start = perf.now();
+        var ret = reactionFunction.apply(this, arguments);
+        var end = perf.now();
+        _digestEvents.push({
+          eventType: 'scope:reaction',
+          id: this.$id,
+          watch: watchStr,
+          time: end - start
+        });
+        return ret;
```
Comment (erwinmombay) [inline src/modules/scopes.js]: this function and the previous functions before it that are getting assigned to arguments seem to only have very small differences (the ret and the eventType?). would it be possible to refactor it and move them into a function or maybe create a decorator? let me know if this isn't doable. (still trying to get a feel for the code etc)

Comment (SomeKittens) [inline src/modules/scopes.js]: I did look into merging them but there's a few subtle differences (how `ret` is calculated, for instance) that make it easier to have them split out.

Comment (erwinmombay) [inline src/modules/scopes.js]: i see. something like this(below) wouldn't work? (or are there perf implications, or would it affect the timer maybe?)  

    function someHelper(eventType, id, watStr, getRet) {
      return function() {
        var start = perf.now();
        var ret = getRet();
        var end = perf.now();
        _digestEvents.push({
          eventType: 'scope:reaction',
          id: this.$id,
          watch: watchStr,
          time: end - start
        });
      };
    }

    arguments[0] = someHelper('scope:watch', id, watchStr, (function(args) {
      return watchExpression.apply(this, args);
    }).bind(this, arguments));

    arguments[0] = someHelper('scope:watch', scopeId, watchStr, (function(watchExpression) {
      return this.$eval(watchExpression);
    }).bind(thatScope, watchExpression));

    arguments[1] = someHelper('scope:reaction', this.$id, watchStr, (function(args) {
      return reactionFunction.apply(this, args);
    }).bind(this, arguments));

Comment (SomeKittens) [inline src/modules/scopes.js]: There's a few tweaks needed in your code (`if` statement and such) but I think that having clear intent trumps having fewer lines of code.  With the code you've written above, readers will need to look in two places to figure out what a particular action does (as opposed to now, where they only need one).

Comment (erwinmombay) [inline src/modules/scopes.js]: sounds good :+1:

**Comment by btford** [comment]
Landed as 3827a76d15b014136293f1f3a35b32e8e2c852e5.

**Pull request closed without merging**